examples/benchmark/datagen.py

```diff
@@ -0,0 +1,345 @@
+            if new_total > size_threshold:
+                yield group
+                group = [item]
+                total = item_size
+            else:
+                group.append(item)
+                total = new_total
+        if len(group) > 0:
+            yield group
+
+    @staticmethod
+    def get_sample_size(item: tuple[str, io.BytesIO]) -> int:
+        nbytes = item[1].getbuffer().nbytes
+        return nbytes
+
+    @staticmethod
+    def tar_samples(
+        samples: list[tuple[str, io.BytesIO]], ctr: ThreadSafeCounter
+    ) -> tuple[str, io.BytesIO]:
+        tar_fileobj = io.BytesIO()
+        with tarfile.open(fileobj=tar_fileobj, mode="w|") as tar:
+            for name, sample in samples:
+                tf = tarfile.TarInfo(name=name)
+                tf.mtime = time.time()
+                tf.size = sample.getbuffer().nbytes
+
+                tar.addfile(tf, sample)
+        tar_fileobj.seek(0)
+
+        return f"shard_{ctr.get_and_inc()}.tar", tar_fileobj
+
+    @staticmethod
+    def parse_resolution(
+        ctx: click.Context, param: str, value: str
+    ) -> tuple[int, int] | None:
+        if value is not None:
+            width, _, height = value.replace(" ", "").lower().partition("x")
+            return int(width), int(height)
+
+    @staticmethod
+    def upload_to_s3(region: str, data: io.BytesIO, bucket: str, key: str):
+        click.echo(f"Uploading to {key=}")
+        s3_client = boto3.client("s3", region_name=region)
+        s3_client.upload_fileobj(data, bucket, key)
+
+    @staticmethod
+    def parse_human_readable_bytes(
+        ctx: click.Context, param: str, value: str
+    ) -> float | None:
+        if value is not None:
+            return pr.Float(value.rstrip("b").rstrip("B"))
+
+    @classmethod
+    def write_dataset_config(cls, disambiguator: str, dataset_cfg: dict[str, Any]):
+        cfg_path = f"./configuration/dataset/{disambiguator}.yaml"
+        with open(cfg_path, "w") as outfile:
+            yaml.dump(dataset_cfg, outfile, default_flow_style=False)
+            click.echo(f"Dataset Configuration created at: {cfg_path}")
+
+
+Sentinel = NewType("Sentinel", None)
+
+
+class ThreadSafeIterator:
+    """Takes an iterator/generator and makes it thread-safe by
+    serializing call to the `next` method of given iterator/generator.
+    """
+
+    def __init__(self, it):
+        self.it = it
+        self.lock = Lock()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        with self.lock:
+            return self.it.__next__()
+
+
+def build_pipeline(
+    num_samples: float, resolution: tuple[int, int], shard_size: float
+) -> Iterator[tuple[str, io.BytesIO]]:
+    # define the pipeline to generate the dataset
+    # TODO: parameterize the data generator to allow for creating other kinds of datasets(eg: text).
+    sample_generator = ImageGenerator(*resolution)
+    pipeline: Iterator[tuple[str, io.BytesIO]] = sample_generator.create(
```

Comment:
Nit: no need for a generator comprehension here. Just using the range function directly should work